### PR TITLE
Adding a cancelled field to tell if a cancellable task is cancelled 

### PR DESCRIPTION
**Pull request by meghasaik**
Signed-off-by: Megha Sai Kavikondala <kavmegha@amazon.com>

### Description
A cancelled field is added which tells if the cancellable task is cancelled. At some points, even if the cancellable task is cancelled it still runs, so to show that the task is cancelled, a cancelled field is shown in the tasks API, `cancelled:true` .
 
### Issues Resolved

https://github.com/opensearch-project/OpenSearch/issues/520 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/tasks/TaskInfo.java [hunk 1]
```diff
@@ -64,6 +65,9 @@
  * snapshot information about currently running tasks.
  */
 public final class TaskInfo implements Writeable, ToXContentFragment {
+
+    static final String INCLUDE_CANCELLED_PARAM = "include_cancelled";
```
Comment (dblock) [inline server/src/main/java/org/opensearch/tasks/TaskInfo.java:69]: I am not super familiar with this code, why is this necessary?

Comment (meghasaik) [reply server/src/main/java/org/opensearch/tasks/TaskInfo.java:69]: We don’t want this field to be shown if task is completed and also as an entry in the task index as we are not mapping this dynamically.

Comment (dblock) [reply server/src/main/java/org/opensearch/tasks/TaskInfo.java:69]: Where else are we doing something similar?

File: server/src/test/java/org/opensearch/action/admin/cluster/node/tasks/TaskTests.java [hunk 2]
```diff
@@ -70,6 +72,11 @@ public void testTaskInfoToString() {
         assertEquals(((Number) map.get("start_time_in_millis")).longValue(), startTime);
         assertEquals(((Number) map.get("running_time_in_nanos")).longValue(), runningTime);
         assertEquals(map.get("cancellable"), cancellable);
+        if (cancellable) {
```
Comment (dblock) [inline server/src/test/java/org/opensearch/action/admin/cluster/node/tasks/TaskTests.java:75]: Having `if` inside tests is an antipattern, it causes tests to start randomly failing. This test will randomly flip cancellable and not and sometimes test the functionality, and sometimes not. Split into three tests: the existing one that tests the kitchen sink, one that only tests cancellable and the other than only tests non-cancellable options.

File: server/src/main/java/org/opensearch/tasks/TaskInfo.java (hunk 1 above)
Comment (dblock) [reply server/src/main/java/org/opensearch/tasks/TaskInfo.java:69]: I also don't see any new tests that utilize this behavior.

Comment (andrross) [reply server/src/main/java/org/opensearch/tasks/TaskInfo.java:69]: Why don't we want this field to be shown if the task is completed?

If being completed it intended to be mutually exclusive from being cancelled, did you consider adding something like a `state` field where the value could be one of `RUNNING`, `CANCELLED` or `COMPLETED`?

Comment (meghasaik) [reply server/src/main/java/org/opensearch/tasks/TaskInfo.java:69]: As the cancelled field tells about the task being cancelled or not, having it when the task is completed felt unnecessary. If the task is completed it means that it has not been cancelled and as mentioned in the issue, if the task is cancelled it will eventually abort/fail. I think the cancelled field does the job of the state field that you mentioned as it shows the `false` = task is running and `true` = task is cancelled.

File: server/src/main/java/org/opensearch/tasks/Task.java [hunk 3]
```diff
@@ -130,6 +130,7 @@ protected final TaskInfo taskInfo(String localNodeId, String description, Status
             startTime,
             System.nanoTime() - startTimeNanos,
             this instanceof CancellableTask,
+            this instanceof CancellableTask && ((CancellableTask) this).isCancelled(),
```
[the remark this answers is not in the document]
Comment (andrross) [reply server/src/main/java/org/opensearch/tasks/Task.java:133]: Agreed. It seems to defeat the purpose of inheritance to have a parent class check at runtime whether it is an instance of a child class.

File: server/src/main/java/org/opensearch/tasks/TaskInfo.java (hunk 1 above)
Comment (andrross) [reply server/src/main/java/org/opensearch/tasks/TaskInfo.java:69]: > it shows the false = task is running and true = task is cancelled

The logic here implicitly has a third state of the cancelled boolean, which is "not present = task is complete". If there is a strong reason to have that third state then using a type that can explicitly represent three states is a better way to go. However, I'd consider doing the simpler approach of always including the cancelled boolean, even on complete tasks when it will be set to false.

Comment (meghasaik) [reply server/src/main/java/org/opensearch/tasks/TaskInfo.java]: I have added the cancelled field to be shown as a field when the task is completed.

File: server/src/main/java/org/opensearch/tasks/TaskInfo.java [hunk 4]
```diff
@@ -119,6 +129,12 @@ public TaskInfo(StreamInput in) throws IOException {
         startTime = in.readLong();
         runningTimeNanos = in.readLong();
         cancellable = in.readBoolean();
+        if (in.getVersion().onOrAfter(Version.V_2_0_0)) {
+            cancelled = in.readBoolean();
+        } else {
+            cancelled = false;
+        }
+        assert cancellable || cancelled == false : "uncancellable task cannot be cancelled";
```
[the remark this answers is not in the document]
Comment (meghasaik) [reply server/src/main/java/org/opensearch/tasks/TaskInfo.java:137]: This might trigger in those cases where we are trying to cancel a task which cannot be cancelled when cancelled is true and cancellable is false. But as you mentioned we can throw a IllegalArgumentException instead of the assert error.

**Comment by meghasaik** [comment]
Closing this PR out as I was getting merge conflicts.
https://github.com/opensearch-project/OpenSearch/pull/1732
